src/b-splines/indexing.jl

```diff
@@ -98,11 +98,14 @@ end
     :(gradient!(g, itp, $(args...)))
 end
 
-@generated function gradient{T,N}(itp::AbstractInterpolation{T,N}, xs...)
-    n = count_interp_dims(itp, N)
-    Tg = promote_type(T, [x <: AbstractArray ? eltype(x) : x for x in xs]...)
-    xargs = [:(xs[$d]) for d in 1:length(xs)]
-    :(gradient!(Array($Tg,$n), itp, $(xargs...)))
+# @eval uglyness required for disambiguation with method in Base
+for R in [:Real, :Any]
+    @eval @generated function gradient{T,N}(itp::AbstractInterpolation{T,N}, xs::$R...)
```

Comment:
Using `@eval` here just to avoid code duplication seems to obscure the code more than it helps. It's probably better just to duplicate the method definition - once with `xs::Real...` and once with `xs...`. It's of course entirely possible to lift out the actual implementation into a private helper, i.e. 

```
@generated function gradient(itp::AbstractInterpolation, xs::Real...) = _gradient(itp, xs...)
@generated function gradient(itp::AbstractInterpolation, xs::Any...) = _gradient(itp, xs...)
@generated function _gradient{T,N}(itp::AbstractInterpolation{T,N}(itp::AbstractInterpolation{T,N}, xs...)
    #current implementation
end
```